client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,121 @@
+  };
+
+  declare activeClass: string;
+  declare activeValue: string;
+  declare durationSecondsValue: number;
+  declare labelTarget: HTMLElement;
+  declare hasActiveClass: boolean;
+  declare hasLabelTarget: boolean;
+  declare originalText: string;
+
+  static afterLoad(identifier: string, application: Application) {
+    const { controllerAttribute } = application.schema;
+    const { actionAttribute } = application.schema;
+
+    const updateLoader = () => {
+      const em = document.querySelector('em') as HTMLElement;
+      document
+        .querySelectorAll(
+          `.button-longrunning:not([data-controller~='${identifier}'])`,
+        )
+        .forEach((button) => {
+          button.setAttribute(controllerAttribute, identifier);
+          button.setAttribute(actionAttribute, 'w-loader#activate');
+
+          if (button.getAttribute('data-clicked-text')) {
+            button.setAttribute(
+              'data-w-loader-active-value',
+              `${button.getAttribute('data-clicked-text')}`,
+            );
+            button.removeAttribute('data-clicked-text');
+          }
+
+          if (em.matches('button em')) {
+            em.setAttribute('data-w-loader-target', 'label');
+          }
+
+          button.setAttribute('data-w-loader-duration-seconds-value', '10');
+        });
+    };
+
+    document.addEventListener('DOMContentLoaded', updateLoader);
+  }
+
+  activate() {
+    this.originalText = this.hasLabelTarget
+      ? this.labelTarget.innerText
+      : (this.element.textContent as string);
+
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      return;
+    }
+
+    // set time out
+    this.setTimeout();
+  }
+
+  cancelSpinner(originalText: string) {
+    this.element.removeAttribute('disabled');
+    this.element.classList.remove(
+      this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
+    );
+
+    if (this.activeValue && this.hasLabelTarget) {
+      this.labelTarget.innerText = originalText;
+    }
+  }
+
+  setTimeout() {
+    const timeout = setTimeout(() => {
+      if (this.activeValue && this.hasLabelTarget === true) {
+        // Sets the targetElement innerText to the activeValue
+        this.labelTarget.innerText = this.activeValue;
+      }
+
+      // Disabling button must be done last: disabled buttons can't be
+      // modified in the normal way, it would seem.
+      this.element.classList.add(
+        this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
+      );
+      this.element.setAttribute('disabled', '');
+
+      setTimeout(() => {
+        this.cancelSpinner(this.originalText);
+      }, this.durationSecondsValue * 1000);
+
+      clearTimeout(timeout);
+    }, 10);
+  }
+}
```

Comment:
I am going to recommend a different approach here;

1. Avoid writing a method called `setTimeout` - this can be confusing as there is also a global `setTimeout` and could easily cause bugs.
2. Instead, set up the timer in the `activate` method, doing the pre-work (setting the attributes like disabled / label stuff) BEFORE the timer starts. This was the other main problem with the previous approach, you were only setting the disabled state after the timer ran by the looks of things.
3. Now, rename the `cancelSpinner` to a simpler method either `cancel` or `reset` - this avoids the assumption that there is a spinner in the first place and is a bit easier to read.
4. The other thing to consider is what happens if the controller disconnects before the timer finishes, while this would probably not be a huge problem (just a console error), it is good to always 'clean up' your timers in disconnect.
5. Small note - while it will work the same, best to use the `window.setTimeout` - it appears that TypeScript assumes the code is running in Node and has a different type if not done this way.

All of this produces this revised approach...

```ts

 /// other static stuff ...
   // small nitpick - remember to alpha sort these - it makes it easier to read.
  declare activeClass: string;
  declare activeValue: string;
  declare durationSecondsValue: number;
  declare hasActiveClass: boolean;
  declare hasLabelTarget: boolean;
  declare labelTarget: HTMLElement;
  declare originalText: string;
  timer?: number; // this is intentionally not using `declare` and is optional (e.g. it may not exist on the instance)

  // afterLoad...
  activate() {
    // this syntax is a bit cluncky but instead of forcing the type into a string, think about how to always have a string, even if textContent is not present, using `||` will allow this whole thing to resolve to an empty string if no label can be found.
    this.originalText =
      (this.hasLabelTarget
        ? this.labelTarget.textContent
        : this.element.textContent) || '';

    // If client-side validation is active on this form, and is going to block submission of the
    // form, don't activate the spinner
    const form = this.element.closest('form');

    if (
      form &&
      form.checkValidity &&
      !form.noValidate &&
      !form.checkValidity()
    ) {
      return;
    }

    // big change here - instead of a method setTimeout we now just 'set up' our active state straight away
    // if this causes a bug (maybe the behaviour gets blocked) we may need to delay this but we can do it inside a promise if needed to delay until the next tick
   // this is a bit tricky to understand but basically, when you disable a button, that disabling happens before the click fully 'completes'
  // to avoid that, we 'defer' the behaviour to after the click has completed, there are multiple ways to do this (a timeout with 0ms or a promise that resolves straight away. Here we are using the first timeout to just 'defer' the initial set up. Similar to what you had originally but just a bit more readable in the same method.

    window.setTimeout(() => {
      if (this.activeValue && this.hasLabelTarget) {
        this.labelTarget.textContent = this.activeValue;
      }

      this.element.classList.add(
        this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
      );

      this.element.setAttribute('disabled', '');
    });

    const durationMs = this.durationSecondsValue * 1000;
    this.timer = window.setTimeout(() => this.reset(), durationMs);
  }

  reset() {
    const originalText = this.originalText;

    this.element.removeAttribute('disabled');

    this.element.classList.remove(
      this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
    );

    if (this.activeValue && this.hasLabelTarget && originalText) {
      this.labelTarget.textContent = originalText;
    }
  }

  disconnect(): void {
    if (this.timer) {
      clearTimeout(this.timer);
    }
  }
```